src/main/scala/cosas/fns.scala

```diff
@@ -47,4 +47,140 @@ object fns {
+  trait AnyDepFn2 extends AnyDepFn { type In1; type In2 }
+  trait DepFn2[I1,I2,O] extends AnyDepFn2 { type In1 = I1; type In2 = I2; type Out = O }
+  // TODO moar arities
+  // trait AnyDepFn3 extends AnyDepFn { type In1; type In2 }
+
+  trait AnyDepFn1Composition extends AnyDepFn1 {
+
+    type First <: AnyDepFn1
+    val first: First
+    type Second <: AnyDepFn1 { type In1 >: First#Out }
+    val second: Second
+
+    type In1 <: First#In1
+    type Out <: Second#Out
+  }
+
+  case class Composition[F <: AnyDepFn1, S <: AnyDepFn1 { type In1 >: F#Out }](val first: F, val second: S) extends AnyDepFn1Composition {
+
+    type First = F
+    type In1 = First#In1
+    type Second = S
+    type Out = Second#Out
+
+    implicit def appForMe[
+      X1 <: F#In1,
+      M <: S#In1,
+      O <: S#Out
+    ](implicit
+      appF: App1[F,X1] { type Out = M },
+      appS: App1[S,M] { type Out = O }
+    ): App1[this.type,X1] { type Out = O } = app1 { x1: X1 => appS(appF(x1)) }
+  }
+
+  implicit final def depFn1Ops[DF <: AnyDepFn1](df: DF): DepFn1Ops[DF] =
+    DepFn1Ops(df)
+  case class DepFn1Ops[DF <: AnyDepFn1](val df: DF) extends AnyVal {
+
+    def at[X <: DF#In1, Y <: DF#Out](f: X => Y): app1[DF,X,Y] =
+      app1[DF,X,Y](f)
+  }
+
+  implicit final def depFn2Ops[DF <: AnyDepFn2](df: DF): DepFn2Ops[DF] =
+    DepFn2Ops(df)
+  case class DepFn2Ops[DF <: AnyDepFn2](val df: DF) extends AnyVal {
+
+    def at[X1 <: DF#In1, X2 <: DF#In2, Y <: DF#Out](f: (X1,X2) => Y): app2[DF,X1,X2,Y] =
+      app2(f)
+  }
+
+  /* dependent function application machinery. These are to be thought of as the building blocks for terms of a dependent function type. */
+  trait AnyAt {
+
+    type DepFn <: AnyDepFn
+    type Out <: DepFn#Out
+  }
+  trait At[P0 <: AnyDepFn] extends AnyAt {  type DepFn = P0  }
+
+  trait At0[DF0 <: AnyDepFn0] extends At[DF0] {
+
+    def apply: Out
+  }
+
+  trait At1[DF1 <: AnyDepFn1] extends At[DF1] {
+
+    type In1 <: DepFn#In1
+
+    def apply(in: In1): Out
+  }
+
+  trait At2[DF2 <: AnyDepFn2] extends At[DF2] {
+
+    type In1 <: DepFn#In1
+    type In2 <: DepFn#In2
+
+    def apply(in1: In1, in2: In2): Out
+  }
+
+  // TODO case-sensitive blahblahblah
+  trait App1[DF <: AnyDepFn1, I <: DF#In1] extends At1[DF] { type In1 = I }
+  case class app1[DF <: AnyDepFn1, I <: DF#In1, O <: DF#Out](val does: I => O) extends App1[DF,I] {
+
+    type Out = O
+
+    final def apply(in: In1): Out =
+      does(in)
+  }
+
+  implicit def depFn1ApplyOps[DF <: AnyDepFn1, X1 <: DF#In1](df: DF): DepFn1ApplyOps[DF,X1] =
+    DepFn1ApplyOps(df)
+  case class DepFn1ApplyOps[DF <: AnyDepFn1, X1 <: DF#In1](val df: DF) extends AnyVal {
+
+    def apply[O](x1: X1)(implicit app: App1[DF,X1] { type Out = O }): O =
+      app(x1)
+  }
+
+  implicit def compositionApplyOps[
+    F <: AnyDepFn1, S <: AnyDepFn1 { type In1 >: F#Out },
+    X1 <: F#In1, M <: F#Out
+  ](comp: Composition[F,S])(implicit
+    appS: App1[S,M]
```

Comment:
this is just a hack that forces `M` to be inferred *before* looking for an `app1[S,M,O]`